cron/pubsub/subscriber.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2021 Security Scorecard Authors
+//
+// Licensed under the Apache License, Vershandlern 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permisshandlerns and
+// limitathandlerns under the License.
+
+package pubsub
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/ossf/scorecard/cron/data"
+	"gocloud.dev/pubsub"
+	_ "gocloud.dev/pubsub/gcppubsub" // Needed to link in GCP drivers.
+	"google.golang.org/protobuf/encoding/protojson"
+)
+
+var (
+	ErrorInReceive  = errors.New("error in Receive")
+	ErrorInParse    = errors.New("error during protojson.Unmarshal")
+	ErrorInShutdown = errors.New("error during shitdown")
+)
+
+type Subscriber interface {
+	SynchronousPull() (*data.ScorecardBatchRequest, error)
+	Ack()
+	Close() error
+}
+
+func CreateSubscriber(ctx context.Context, subscriptionURL string) (Subscriber, error) {
+	subscription, err := pubsub.OpenSubscription(ctx, subscriptionURL)
+	if err != nil {
+		return nil, fmt.Errorf("error during pubsub.OpenSubscription: %w", err)
+	}
+	ret := subscriberImpl{
+		ctx:          ctx,
+		subscription: subscription,
+	}
+	return &ret, nil
+}
+
+type receiver interface {
+	Receive(ctx context.Context) (*pubsub.Message, error)
+	Shutdown(ctx context.Context) error
+}
+
+type subscriberImpl struct {
+	ctx          context.Context
+	subscription receiver
+	msg          *pubsub.Message
+}
+
+func (subscriber *subscriberImpl) SynchronousPull() (*data.ScorecardBatchRequest, error) {
+	msg, err := subscriber.subscription.Receive(subscriber.ctx)
```

Comment:
I recently found out that go-cloud does not allow you to extend the acknowledgement deadline (i.e. the "lease period"), and it does not do it automatically in the background AFAICT. 

This means that the maximum deadline is the 10 minutes allowed by Pub/Sub before the message gets redelivered, which is probably OK if each message is for a single repo, but each message here can potentially be for a large batch of repos which may easily exceed it.

Could we split up the request messages into one-per-repo? There are other benefits too, like better granularity when it comes to retrying requests in case of unexpected transient failures. What are the benefits of batching/sharding requests ourselves?

Reply:
> I recently found out that go-cloud does not allow you to extend the acknowledgement deadline (i.e. the "lease period"), and it does not do it automatically in the background AFAICT.
> 
> This means that the maximum deadline is the 10 minutes allowed by Pub/Sub before the message gets redelivered, which is probably OK if each message is for a single repo, but each message here can potentially be for a large batch of repos which may easily exceed it.
> 

Ah, that's not good. We should have the automatic lease management irrespective of which solution we choose. I'm considering re-writing `subscriber.go` using GCS libraries instead of go-cloud. Wdyt?

> Could we split up the request messages into one-per-repo? There are other benefits too, like better granularity when it comes to retrying requests in case of unexpected transient failures. What are the benefits of batching/sharding requests ourselves?

The benefit is determinism. e.g - (i) `controller` knows `num_shards` beforehand which is useful in figuring out job completion status, (ii) the generated files can have filenames which read `shard-xxx` making them easily listable/queryable. 
The other thing to note here is that, with the current implementation, we can set `shard_size = 1` in the config to achieve what you are looking for without any change to code/implementation.

Reply:
Agreed. `shard_size = 250` is just something I chose for now. We probably need to do some tuning to figure out what works best for us.